### ISSUE TITLE
Can't hover with Apple Pencil on iOS

### ISSUE THREAD

**Issue by neko-kai** [ppy/osu#29594]
### Type

Game behaviour

### Bug description

It's impossible to hover - move the cursor without registering an M1 key press - when using an Apple Pencil. Touching the screen with the pen always presses M1 irrespective of pressure.

Judging by https://github.com/ppy/osu/issues/6401 hover/tap/drag differentiation is supported for Android tablets, but that doesn't seem to be the case for iOS?

### Version

2024.625.2-lazer

### Logs

.

**Comment by Star-Clara** [comment]
But the prerequisite is that you need to have a keyboard💦

**Comment by neko-kai** [comment]
@Star-Clara That allows to hover, but disables dragging or tapping. I'm playing tap+x and have trouble following streams since I have to lift the pen to tap and lose the ability to aim at the same time. Apparently regular tablet (and Android?) users can light press on the pen to hover and increase pressure to tap, so they can aim while streaming.

But seemingly this is not implemented on iOS? Pencil seems to be treated exactly like a regular touch input, even though Apple pencil itself supports pressure and OS APIs expose it too.

**Comment by Jan-Arbus** [comment]
Android pen hover is broken too. So this isn't iOS specific. I think it might be an SDL3 issue, since the mobile framework moved to SDL3 I've noticed that pen hover input isn't processed

**Comment by hwsmm** [comment]
It will get fixed once [this](https://www.github.com/libsdl-org/SDL/issues/10516) is done.

**Comment by frenzibyte** [comment]
@neko-kai I'm having trouble following what your playstyle is. Video recording will immensely help.

**Comment by neko-kai** [comment]
@frenzibyte Let me describe:
By default the stylus is off the screen, same as with pure touch playstyle.

To hit jumps I touch the note with the stylus, then lift it off the screen completely and touch the next note. Same as you would with a touchscreen playstyle.

To hit streams, I touch the note with the stylus, then hit X on the keypad, then lift the pen completely off the screen and aim for the next note. Because the cursor is not moving while the pen is not touching the screen, streaming looks 'staggered' in the replay.

To hit spaced streams, I touch the note, then drag the pen to the next note as if following a slider, then press X on the keypad. Then lift the pen completely off the screen and touch the next note. This is very hard to pull off on high bpm.

How would hover help: in spaced stream case, I could avoid lifting the pen off the screen, hovering to the next note instead.

I can't record today, but hopefully I described it well enough. I can record later if not.

**Comment by frenzibyte** [comment]
There are still questions in my head, mainly: why do you have to lift the pen off the screen in  streams / spaced streams? Does the game not handle input correctly otherwise?

Also you mention the cursor doesn't move when the pen doesn't touch the screen, but this isn't true? There is hovering support on iOS, the cursor *will* move in gameplay if the pen is hovering the screen.

**Comment by neko-kai** [comment]
@frenzibyte
> why do you have to lift the pen off the screen in  streams / spaced streams? Does the game not handle input correctly otherwise?

Yes, that is the exact problem. Osu only produces one input from the pen - drag (mouse click and hold M1 + cursor move), no matter what amount of pressure is applied to the screen. There's no way to move the cursor without also sending M1 click. There's also no way to send M1 click without first unpressing M1 - i.e. lifting the pen from the screen.

> Also you mention the cursor doesn't move when the pen doesn't touch the screen, but this isn't true? There is hovering support on iOS, the cursor *will* move in gameplay if the pen is hovering the screen.

Was that added recently? I did not update the game since the last message. Unless you mean the cursor will still move when registering drag (hold M1 click+move) - yes, it will move, but I still must lift the pen off from the screen ever so slightly in order to de-press M1 and touch the screen again to register the next M1 click.

**Comment by ericshu6** [comment]
I know that this is an old issue, but I’d like to add this just in case someone else is experiencing the same problem: I found that Apple Pencil Pro hover works when High-precision mouse is turned off and osu! is closed and re-opened (I’m on iPad Pro M4).

**Comment by victorbjafet** [comment]
> I know that this is an old issue, but I’d like to add this just in case someone else is experiencing the same problem: I found that Apple Pencil Pro hover works when High-precision mouse is turned off and osu! is closed and re-opened (I’m on iPad Pro M4).

Thank you! This worked. Although, it would of course be nice if this worked out of the box, regardless of the setting, and ideally documented somewhere thats visible to someone playing on an iPad build.
